### PR TITLE
Smarter generate

### PR DESCRIPTION
**Pull request by jesper-friis**
# Description:

## Type of change:

- [ ] Bug fix.
- [ ] New feature.
- [ ] Documentation update.

## Checklist for the reviewer:

This checklist should be used as a help for the reviewer.

- [ ] Is the change limited to one issue?
- [ ] Does this PR close the issue?
- [ ] Is the code easy to read and understand, including clearly named variables?
- [ ] Do all new feature have an accompanying new test?
- [ ] Has the documentation been updated as necessary?

### CODE REVIEW

File: oteapi_dlite/strategies/generate.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        # Can we savely assume that all strategies in a pipeline will be
+        # executed in the same Python interpreter?  If not, we should write
+        # the collection to a storage, such that it can be shared with the
+        # other strategies.
+
+        update_collection(coll)
+        return DLiteSessionUpdate(collection_id=coll.uuid)
+
+
+# DLiteStorageConfig.update_forward_refs()
```
Comment (daniel-sintef) [inline oteapi_dlite/strategies/generate.py:170]: Comment should be removed if unused

**Comment by daniel-sintef** [comment]
I noticed that the CI doesn't seem to run any of the test_generate scripts you added? First, can you confirm by looking at the CI report, then confirm by running on your local machine? (I tried running on mine but the installation seems complex [I am getting a large number of unrelated crashes] and I won't have time to figure it out by this afternoon)

**Comment by francescalb** [comment]
> I noticed that the CI doesn't seem to run any of the test_generate scripts you added? First, can you confirm by looking at the CI report, then confirm by running on your local machine? (I tried running on mine but the installation seems complex [I am getting a large number of unrelated crashes] and I won't have time to figure it out by this afternoon)

Just a comment on this, I also thought the same, since modules tested to no appear in the generated list by pytest. However, pytest does run them, and hard crashes everything if one of the tests fail. Instead, if they are written as separate methods all of them are tested and failures are printed for all the fail. At least this is the case on my system.

File: oteapi_dlite/strategies/generate.py (hunk 1 above)
Comment (jesper-friis) [reply oteapi_dlite/strategies/generate.py]: fixed

**Comment by jesper-friis** [comment]
> > I noticed that the CI doesn't seem to run any of the test_generate scripts you added? First, can you confirm by looking at the CI report, then confirm by running on your local machine? (I tried running on mine but the installation seems complex [I am getting a large number of unrelated crashes] and I won't have time to figure it out by this afternoon)
> 
> Just a comment on this, I also thought the same, since modules tested to no appear in the generated list by pytest. However, pytest does run them, and hard crashes everything if one of the tests fail. Instead, if they are written as separate methods all of them are tested and failures are printed for all the fail. At least this is the case on my system.

Yes, that is also my experience. Test scripts have a file-level granularity while test functions have function-level granularity.
Hence, with test scripts only one failure will be reported per script, while for test functions max one failure will be reported per function. The same is the case for test separation (one test influencing another).

**Comment by francescalb** [comment]
> > > I noticed that the CI doesn't seem to run any of the test_generate scripts you added? First, can you confirm by looking at the CI report, then confirm by running on your local machine? (I tried running on mine but the installation seems complex [I am getting a large number of unrelated crashes] and I won't have time to figure it out by this afternoon)
> > 
> > 
> > Just a comment on this, I also thought the same, since modules tested to no appear in the generated list by pytest. However, pytest does run them, and hard crashes everything if one of the tests fail. Instead, if they are written as separate methods all of them are tested and failures are printed for all the fail. At least this is the case on my system.
> 
> Yes, that is also my experience. Test scripts have a file-level granularity while test functions have function-level granularity. Hence, with test scripts only one failure will be reported per script, while for test functions max one failure will be reported per function. The same is the case for test separation.
> 
> In either case will pytest run all the tests.

Not exactly. My point is that if a test_script fails (e.g. test_generate.py) pytest stops and reports only the error for that one. It does not continute checking the other scripts or functions as well. Instead, when a function fails, it will will still go through all functions and report on all of them. I much prefer having a ffull report with separate test_functions.

**Review by daniel-sintef — APPROVED** [review]

**Pull request merged**